File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> results;
        vector<int> result;
        int n = nums.size();
        bool used[n];
        
        for(int i = 0; i < n; i++) used[i] = false;
        
        dfs(n, used, nums, result, results);
        return results;
    }
    
    void dfs(int n, bool used[], vector<int>& nums, vector<int>& result, vector<vector<int>>& results)
    {
        if(result.size() == n)
        {
            results.push_back(result);
            return;
        }
        
        for(int i = 0; i < n; i++)
        {
            if(used[i]) continue;
            result.push_back(nums[i]);
            used[i] = true;
            dfs(n, used, nums, result, results);
            used[i] = false;
            result.pop_back();
        }
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def backtracking(self, nums, result, results, used):
        #base case
        if len(result) == len(nums):
            results.append(result[:])
            return
        
        #recursive case
        for i in range(len(nums)):
            num = nums[i]
            
            if used[i]:
                continue
            
            result.append(num)
            used[i] = True
            self.backtracking(nums, result, results, used)
            result.pop()
            used[i] = False
        
    
    
    def permute(self, nums: List[int]) -> List[List[int]]:
        results = [] #store all the orders
        result = [] #store the current order we are building
        used = [False for _ in range(len(nums))] #store which number has been used
        
        self.backtracking(nums, result, results, used)
        
        return results